In [2]:
import numpy as np
import pandas as pd

# Task 1

In [3]:
space_probs = pd.read_csv('space_prob.csv')
space_probs

Space (ft3)  Probability
0             1         0.08
1             2         0.12
2             3         0.15
3             4         0.12
4             5         0.08
5             6         0.06
6             7         0.05
7             8         0.05
8             9         0.05
9            10         0.04
10           11         0.04
11           12         0.03
12           13         0.03
13           14         0.03
14           15         0.02
15           16         0.02
16           17         0.02
17           18         0.01

Then we are given 4-week period demand share.

In [4]:
demand_share_4_weeks = pd.read_csv('4-week_demand_share.csv').T
demand_share_4_weeks.columns = ['mean_demand_share_4_weeks']
demand_share_4_weeks.reset_index(inplace=True)
demand_share_4_weeks.drop('index', axis=1, inplace=True)
demand_share_4_weeks

mean_demand_share_4_weeks
0                        0.06
1                        0.05
2                        0.06
3                        0.07
4                        0.09
5                        0.06
6                        0.04
7                        0.10
8                        0.10
9                        0.06
10                       0.08
11                       0.10
12                       0.13

In [5]:
Cv_4_week = 0.3
np.random.seed(42)

demand_share_4_weeks['sigma_4_weeks'] = Cv_4_week * demand_share_4_weeks['mean_demand_share_4_weeks']
for i in range(len(demand_share_4_weeks)):
    demand_share_4_weeks.loc[i, 'net_4_weeks_demand'] = np.random.normal(demand_share_4_weeks.loc[i, 'mean_demand_share_4_weeks'], demand_share_4_weeks.loc[i, 'sigma_4_weeks'])
demand_share_4_weeks

mean_demand_share_4_weeks  sigma_4_weeks  net_4_weeks_demand
0                        0.06          0.018            0.068941
1                        0.05          0.015            0.047926
2                        0.06          0.018            0.071658
3                        0.07          0.021            0.101984
4                        0.09          0.027            0.083678
5                        0.06          0.018            0.055786
6                        0.04          0.012            0.058951
7                        0.10          0.030            0.123023
8                        0.10          0.030            0.085916
9                        0.06          0.018            0.069766
10                       0.08          0.024            0.068878
11                       0.10          0.030            0.086028
12                       0.13          0.039            0.139437

In [6]:
demand_share_4_weeks['net_4_weeks_demand'] = demand_share_4_weeks['net_4_weeks_demand'] / demand_share_4_weeks['net_4_weeks_demand'].sum(axis=0)
demand_share_4_weeks

mean_demand_share_4_weeks  sigma_4_weeks  net_4_weeks_demand
0                        0.06          0.018            0.064918
1                        0.05          0.015            0.045129
2                        0.06          0.018            0.067477
3                        0.07          0.021            0.096032
4                        0.09          0.027            0.078795
5                        0.06          0.018            0.052530
6                        0.04          0.012            0.055511
7                        0.10          0.030            0.115844
8                        0.10          0.030            0.080902
9                        0.06          0.018            0.065695
10                       0.08          0.024            0.064859
11                       0.10          0.030            0.081008
12                       0.13          0.039            0.131300

In [7]:
daily_demand_share = pd.read_csv('demand_share_per_day.csv').T
daily_demand_share.columns = ['mean_demand_share_per_day']
daily_demand_share.reset_index(inplace=True)
daily_demand_share.drop('index', axis=1, inplace=True)
daily_demand_share

mean_demand_share_per_day
0                       0.05
1                       0.10
2                       0.15
3                       0.15
4                       0.18
5                       0.22
6                       0.15

In [8]:
Cv_per_day = 0.2
np.random.seed(42)

daily_demand_share['sigma_per_day'] = Cv_per_day * daily_demand_share['mean_demand_share_per_day']
for i in range(len(daily_demand_share)):
    daily_demand_share.loc[i, 'net_daily_demand'] = np.random.normal(daily_demand_share.loc[i, 'mean_demand_share_per_day'], daily_demand_share.loc[i, 'sigma_per_day'])
daily_demand_share

mean_demand_share_per_day  sigma_per_day  net_daily_demand
0                       0.05          0.010          0.054967
1                       0.10          0.020          0.097235
2                       0.15          0.030          0.169431
3                       0.15          0.030          0.195691
4                       0.18          0.036          0.171570
5                       0.22          0.044          0.209698
6                       0.15          0.030          0.197376

In [9]:
daily_demand_share['net_daily_demand'] = daily_demand_share['net_daily_demand'] / daily_demand_share['net_daily_demand'].sum(axis=0)
daily_demand_share

mean_demand_share_per_day  sigma_per_day  net_daily_demand
0                       0.05          0.010          0.050154
1                       0.10          0.020          0.088720
2                       0.15          0.030          0.154594
3                       0.15          0.030          0.178555
4                       0.18          0.036          0.156547
5                       0.22          0.044          0.191336
6                       0.15          0.030          0.180093

In [10]:
daily_demand_share_4_weeks = daily_demand_share.copy()
for i in range(3):
    daily_demand_share_4_weeks = pd.concat([daily_demand_share_4_weeks, daily_demand_share], ignore_index=True, axis=0)
daily_demand_share_4_weeks

mean_demand_share_per_day  sigma_per_day  net_daily_demand
0                        0.05          0.010          0.050154
1                        0.10          0.020          0.088720
2                        0.15          0.030          0.154594
3                        0.15          0.030          0.178555
4                        0.18          0.036          0.156547
5                        0.22          0.044          0.191336
6                        0.15          0.030          0.180093
7                        0.05          0.010          0.050154
8                        0.10          0.020          0.088720
9                        0.15          0.030          0.154594
10                       0.15          0.030          0.178555
11                       0.18          0.036          0.156547
12                       0.22          0.044          0.191336
13                       0.15          0.030          0.180093
14                       0.05          0.010          0.050154
15                       0.10          0.020          0.088720
16                       0.15          0.030          0.154594
17                       0.15          0.030          0.178555
18                       0.18          0.036          0.156547
19                       0.22          0.044          0.191336
20                       0.15          0.030          0.180093
21                       0.05          0.010          0.050154
22                       0.10          0.020          0.088720
23                       0.15          0.030          0.154594
24                       0.15          0.030          0.178555
25                       0.18          0.036          0.156547
26                       0.22          0.044          0.191336
27                       0.15          0.030          0.180093

In [11]:
four_weekly_facs = [f for f in demand_share_4_weeks['net_4_weeks_demand'] for _ in range(28)]
len(four_weekly_facs)

364

In [12]:
yearly_probs = pd.DataFrame({'four_weekly_facs': four_weekly_facs, 'daily_demand_share_4_weeks': list(daily_demand_share_4_weeks['net_daily_demand'])*13})
yearly_probs['yearly_prob'] = yearly_probs['four_weekly_facs'] * yearly_probs['daily_demand_share_4_weeks']
yearly_probs

four_weekly_facs  daily_demand_share_4_weeks  yearly_prob
0            0.064918                    0.050154     0.003256
1            0.064918                    0.088720     0.005760
2            0.064918                    0.154594     0.010036
3            0.064918                    0.178555     0.011591
4            0.064918                    0.156547     0.010163
..                ...                         ...          ...
359          0.131300                    0.154594     0.020298
360          0.131300                    0.178555     0.023444
361          0.131300                    0.156547     0.020555
362          0.131300                    0.191336     0.025122
363          0.131300                    0.180093     0.023646

[364 rows x 3 columns]

In [13]:
yearly_probs['yearly_prob'] = yearly_probs['yearly_prob'] / sum(yearly_probs['yearly_prob'])
yearly_probs

four_weekly_facs  daily_demand_share_4_weeks  yearly_prob
0            0.064918                    0.050154     0.000814
1            0.064918                    0.088720     0.001440
2            0.064918                    0.154594     0.002509
3            0.064918                    0.178555     0.002898
4            0.064918                    0.156547     0.002541
..                ...                         ...          ...
359          0.131300                    0.154594     0.005075
360          0.131300                    0.178555     0.005861
361          0.131300                    0.156547     0.005139
362          0.131300                    0.191336     0.006281
363          0.131300                    0.180093     0.005912

[364 rows x 3 columns]

In [14]:
yearly_probs['yearly_prob'].max()

0.00628058992997273

In [19]:
annaul_demand_forecast = pd.read_csv("annual demand forecast.csv")
annaul_demand_forecast = annaul_demand_forecast.rename(columns={'1': "Year_1", '2': "Year_2", '3': "Year_3", '4': "Year_4", '5': "Year_5"})
annaul_demand_forecast['Scenario'] = [0,1,2,3,4]
annaul_demand_forecast

Scenario  Probability  Year_1  Year_2  Year_3  Year_4  Year_5
0         0         0.30   80000   90000  100000  105000  110000
1         1         0.25   60000   65000   70000   80000   90000
2         2         0.20   80000  100000  110000  120000  150000
3         3         0.15   50000   70000   75000   75000   65000
4         4         0.10   50000   60000   60000   60000   60000

In [20]:
df_scenario = annaul_demand_forecast[['Scenario', 'Probability']]
df_scenario

Scenario  Probability
0         0         0.30
1         1         0.25
2         2         0.20
3         3         0.15
4         4         0.10

In [21]:
demand_breakdown = annaul_demand_forecast[['Year_1', 'Year_2', 'Year_3', 'Year_4', 'Year_5']].copy()
demand_breakdown = demand_breakdown.loc[demand_breakdown.index.repeat(364)]
demand_breakdown.reset_index(inplace=True)
#demand_breakdown.drop('index', axis=1, inplace=True)
demand_breakdown.rename(columns={'index': 'scenario'}, inplace=True)
demand_breakdown

scenario  Year_1  Year_2  Year_3  Year_4  Year_5
0            0   80000   90000  100000  105000  110000
1            0   80000   90000  100000  105000  110000
2            0   80000   90000  100000  105000  110000
3            0   80000   90000  100000  105000  110000
4            0   80000   90000  100000  105000  110000
...        ...     ...     ...     ...     ...     ...
1815         4   50000   60000   60000   60000   60000
1816         4   50000   60000   60000   60000   60000
1817         4   50000   60000   60000   60000   60000
1818         4   50000   60000   60000   60000   60000
1819         4   50000   60000   60000   60000   60000

[1820 rows x 6 columns]

In [80]:
# for col in ['Year_1', 'Year_2', 'Year_3', 'Year_4', 'Year_5']:
#     demand_breakdown[col] *= (list(yearly_probs['yearly_prob'])*5)
        
# demand_breakdown

scenario      Year_1      Year_2      Year_3      Year_4      Year_5
0            0   65.117759   73.257479   81.397199   85.467059   89.536919
1            0  115.190757  129.589602  143.988447  151.187869  158.387291
2            0  200.718908  225.808771  250.898635  263.443567  275.988498
3            0  231.828547  260.807116  289.785684  304.274968  318.764253
4            0  203.253885  228.660621  254.067356  266.770724  279.474092
...        ...         ...         ...         ...         ...         ...
1815         4  253.727886  304.473463  304.473463  304.473463  304.473463
1816         4  293.053444  351.664133  351.664133  351.664133  351.664133
1817         4  256.932339  308.318807  308.318807  308.318807  308.318807
1818         4  314.029496  376.835396  376.835396  376.835396  376.835396
1819         4  295.577518  354.693021  354.693021  354.693021  354.693021

[1820 rows x 6 columns]

{0: 0.3, 1: 0.25, 2: 0.2, 3: 0.15, 4: 0.1}

In [100]:
# demand_breakdown['Probability'] = demand_breakdown['scenario'].map(dict(annaul_demand_forecast['Probability']))
demand_breakdown

scenario  Year_1  Year_2  Year_3  Year_4  Year_5
0            0   80000   90000  100000  105000  110000
1            0   80000   90000  100000  105000  110000
2            0   80000   90000  100000  105000  110000
3            0   80000   90000  100000  105000  110000
4            0   80000   90000  100000  105000  110000
...        ...     ...     ...     ...     ...     ...
1815         4   50000   60000   60000   60000   60000
1816         4   50000   60000   60000   60000   60000
1817         4   50000   60000   60000   60000   60000
1818         4   50000   60000   60000   60000   60000
1819         4   50000   60000   60000   60000   60000

[1820 rows x 6 columns]

In [16]:
# demand_breakdown['Year_1'].groupby(by=demand_breakdown['scenario']).sum()

scenario
0    80000.0
1    60000.0
2    80000.0
3    50000.0
4    50000.0
Name: Year_1, dtype: float64

In [22]:
hourly_demand = pd.read_csv('hourly_records.csv')
hourly_demand.drop('Unnamed: 0', axis=1, inplace=True)
hourly_demand = hourly_demand/10000
hourly_demand

0       1       2       3       4       5       6       7       8  \
0   0.0030  0.0020  0.0004  0.0004  0.0004  0.0010  0.0020  0.0024  0.0032   
1   0.0000  0.0024  0.0012  0.0006  0.0004  0.0010  0.0020  0.0024  0.0032   
2   0.0000  0.0000  0.0012  0.0006  0.0003  0.0005  0.0010  0.0012  0.0016   
3   0.0000  0.0000  0.0000  0.0012  0.0006  0.0003  0.0010  0.0012  0.0016   
4   0.0000  0.0000  0.0000  0.0000  0.0036  0.0024  0.0030  0.0036  0.0048   
5   0.0000  0.0000  0.0000  0.0000  0.0000  0.0036  0.0045  0.0045  0.0048   
6   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0060  0.0060  0.0072   
7   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0072  0.0100   
8   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0150   
9   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
10  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
11  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
12  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
13  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
14  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
15  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
16  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
17  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
18  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0010  0.0040  0.0020   
19  0.0005  0.0000  0.0000  0.0000  0.0000  0.0000  0.0015  0.0050  0.0040   
20  0.0008  0.0000  0.0000  0.0000  0.0000  0.0000  0.0020  0.0060  0.0040   
21  0.0010  0.0004  0.0000  0.0000  0.0000  0.0000  0.0012  0.0036  0.0030   
22  0.0020  0.0010  0.0004  0.0000  0.0000  0.0000  0.0012  0.0036  0.0030   
23  0.0030  0.0014  0.0006  0.0000  0.0000  0.0000  0.0016  0.0040  0.0036   

         9  ...      14      15      16      17      18      19      20  \
0   0.0020  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
1   0.0020  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
2   0.0010  ...  0.0002  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
3   0.0010  ...  0.0004  0.0003  0.0000  0.0000  0.0000  0.0000  0.0000   
4   0.0030  ...  0.0012  0.0009  0.0000  0.0000  0.0000  0.0000  0.0000   
5   0.0030  ...  0.0012  0.0009  0.0000  0.0000  0.0000  0.0000  0.0000   
6   0.0044  ...  0.0016  0.0016  0.0020  0.0012  0.0000  0.0000  0.0000   
7   0.0048  ...  0.0016  0.0016  0.0020  0.0016  0.0012  0.0000  0.0000   
8   0.0120  ...  0.0024  0.0024  0.0030  0.0024  0.0018  0.0018  0.0000   
9   0.0120  ...  0.0024  0.0024  0.0036  0.0042  0.0030  0.0018  0.0012   
10  0.0000  ...  0.0030  0.0024  0.0036  0.0042  0.0030  0.0018  0.0012   
11  0.0000  ...  0.0048  0.0036  0.0036  0.0048  0.0030  0.0018  0.0012   
12  0.0000  ...  0.0040  0.0032  0.0032  0.0040  0.0040  0.0012  0.0008   
13  0.0000  ...  0.0108  0.0060  0.0048  0.0060  0.0060  0.0060  0.0036   
14  0.0000  ...  0.0160  0.0144  0.0112  0.0112  0.0080  0.0080  0.0048   
15  0.0000  ...  0.0000  0.0120  0.0120  0.0120  0.0090  0.0066  0.0036   
16  0.0000  ...  0.0000  0.0000  0.0120  0.0150  0.0120  0.0090  0.0048   
17  0.0000  ...  0.0000  0.0000  0.0000  0.0120  0.0180  0.0120  0.0084   
18  0.0010  ...  0.0000  0.0000  0.0000  0.0000  0.0100  0.0120  0.0100   
19  0.0015  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0100  0.0150   
20  0.0012  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0080   
21  0.0008  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
22  0.0008  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   
23  0.0016  ...  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000   

        21      22      23  
0   0.0000  0.0000  0.0000  
1   0.0000  0.0000  0.0000  
2   0.0000  0.0000  0.0000  
3   0.0000  0.0000  0.0000  
4   0.0000  0.0000  0.0000  
5   0.00

In [109]:
df_scenario['Scenario'].values

array([0, 1, 2, 3, 4], dtype=int64)

In [120]:
np.nonzero(hourly_demand.values)[1][13]

1

In [189]:
def simulation(daily_demand_df = yearly_probs, demand_df = demand_breakdown, scenario_df = df_scenario, hourly_demand_df = hourly_demand, size_df = space_probs):
    '''
    With a random probability, a scenario is chosen. Then use the scenario to choose the demand breakdown. So we get 5 years of demand breakdown.Choose a year. 


    '''
    scenario = np.random.choice(scenario_df['Scenario'].values, p=scenario_df['Probability'].values)
    year = np.random.choice([1,2,3,4,5])
    day = np.random.choice(list(range(364)))
    
    size = np.random.choice(size_df['Space (ft3)'].values, p=size_df['Probability'].values)

    non_zero_indices = np.nonzero(hourly_demand_df.values)
    random_index = np.random.choice(len(non_zero_indices[0]))
    deposit = non_zero_indices[0][random_index]
    pickup = non_zero_indices[1][random_index]

    demand = demand_df.loc[demand_df['scenario'] == scenario, f'Year_{year}']* daily_demand_df.loc[day, 'yearly_prob'] * hourly_demand_df.iloc[deposit, pickup]
    demand = demand.reset_index(drop=True)

    demand_instance = [year, day, deposit, pickup, size, demand[day]]

    return demand_instance

In [190]:
demans = simulation()
demans

[2, 112, 6, 11, 2, 0.21340132186724328]

# Task 2

We get demand instance, space availability and a locker bank design. What will the heiristic be

We check if the same size is availiable for this delivery, if not assign it to the smallest size greater than this. If that is also occupied, reject the order.

# Task 3


In [196]:
## Input params ##
min_alpha_avg = 0.99
c_dd1L = 10
c_m = 50
c_s = 50
c_w = 500
d_ds = 
d_s = 
h_max = 15
lamda = 30
m_dist = 7
p_s =
w_max = 20


SyntaxError: invalid syntax (2326288606.py, line 5)

### Assume we have each locker 1.5ft width, 2ft depth and variable height. Max height 9 ft and width is 10.5 ft. Thus $189 ft^3 $
### Thus $ 1 ft^3 $ locker requires height 0.33 ft -- 8% -- 60 units -- 7
#### Thus $ 3 ft^3 $ locker requires height of 1 ft -- 27% -- 200 units -- 3
#### Thus $ 7 ft^3 $ locker requires height of 2.33 ft -- 31% -- 230 units -- 7
#### Thus $ 10 ft^3 $ locker requires height of 3.33 ft -- 14% -- 105 units -- 4
#### Thus $ 15 ft^3 $ locker requires height of 5 ft -- 15% -- 115 units -- 3
#### Thus $ 18 ft^3 $ locker requires height of 6 ft -- 5% -- 38 units -- 2
#### One interface of 1ft height will be the interactive module that controls access to the lockers

#### We can fit 12 columns of 1.5 ft width lockers in 18 ft total. 
#### Max we'll ever have is 750 packages in a day. 

In [32]:
locker_size_dict = {1:7,3:3,7:7,10:4, 15:3, 18:2}


In [35]:
def daily_simulation(day, daily_demand_df = yearly_probs, demand_df = demand_breakdown, scenario_df = df_scenario, hourly_demand_df = hourly_demand, size_df = space_probs):
    '''
    With a random probability, a scenario is chosen. Then use the scenario to choose the demand breakdown. So we get 5 years of demand breakdown.Choose a year. 


    '''
    scenario = np.random.choice(scenario_df['Scenario'].values, p=scenario_df['Probability'].values)
    year = np.random.choice([1,2,3,4,5])
    # day = np.random.choice(list(range(364)))
    
    size = np.random.choice(size_df['Space (ft3)'].values, p=size_df['Probability'].values)

    # non_zero_indices = np.nonzero(hourly_demand_df.values)
    # random_index = np.random.choice(len(non_zero_indices[0]))
    # deposit = non_zero_indices[0][random_index]
    # pickup = non_zero_indices[1][random_index]

    demand = demand_df.loc[demand_df['scenario'] == scenario, f'Year_{year}']* daily_demand_df.loc[day, 'yearly_prob']
    demand = demand.reset_index(drop=True)
    daily_demand = np.ceil(demand[day])

    demand_yet = 0
    demand_instances = []
    while demand_yet <= daily_demand:
        non_zero_indices = np.nonzero(hourly_demand_df.values)
        random_index = np.random.choice(len(non_zero_indices[0]))
        deposit = non_zero_indices[0][random_index]
        pickup = non_zero_indices[1][random_index]
        size = np.random.choice(size_df['Space (ft3)'].values, p=size_df['Probability'].values)
        demand_order = np.ceil(demand[day]*hourly_demand_df.iloc[deposit, pickup])
        demand_yet += demand_order
        demand_instances.append([deposit, pickup, size,demand_order])


    return demand_instances

In [36]:
orders = daily_simulation(1)
orders.sort(key=lambda x: x[0])
orders

[[0, 6, 8, 1.0],
 [0, 3, 12, 1.0],
 [0, 0, 7, 1.0],
 [0, 9, 5, 1.0],
 [1, 1, 7, 1.0],
 [1, 12, 8, 1.0],
 [1, 10, 4, 1.0],
 [1, 2, 13, 1.0],
 [1, 2, 10, 1.0],
 [1, 10, 4, 1.0],
 [2, 3, 1, 1.0],
 [2, 13, 6, 1.0],
 [2, 13, 4, 1.0],
 [2, 10, 14, 1.0],
 [2, 14, 2, 1.0],
 [2, 13, 3, 1.0],
 [3, 5, 4, 1.0],
 [4, 12, 1, 1.0],
 [4, 13, 14, 1.0],
 [4, 11, 12, 1.0],
 [5, 7, 7, 1.0],
 [5, 10, 3, 1.0],
 [5, 9, 11, 1.0],
 [5, 9, 11, 1.0],
 [6, 8, 4, 1.0],
 [6, 8, 5, 1.0],
 [7, 14, 1, 1.0],
 [7, 13, 2, 1.0],
 [7, 13, 2, 1.0],
 [7, 16, 8, 1.0],
 [7, 14, 6, 1.0],
 [8, 19, 9, 1.0],
 [8, 19, 9, 1.0],
 [8, 9, 4, 2.0],
 [8, 9, 1, 2.0],
 [8, 12, 4, 1.0],
 [8, 12, 3, 1.0],
 [9, 15, 9, 1.0],
 [9, 10, 5, 1.0],
 [9, 15, 11, 1.0],
 [10, 12, 14, 2.0],
 [10, 11, 2, 2.0],
 [10, 19, 17, 1.0],
 [10, 17, 3, 1.0],
 [10, 19, 3, 1.0],
 [11, 19, 3, 1.0],
 [11, 20, 2, 1.0],
 [11, 11, 2, 2.0],
 [11, 13, 15, 1.0],
 [11, 11, 3, 2.0],
 [12, 17, 1, 1.0],
 [12, 13, 3, 2.0],
 [12, 16, 4, 1.0],
 [12, 20, 2, 1.0],
 [12, 22, 4, 1.0],

In [86]:
import copy
def evaluation(day, n_sims = 100):
    locker_sizes = [1,3,7,10,15,18]
    locker_dict = {locker_size: [locker_size_dict[locker_size] for _ in range(24)] for locker_size in locker_sizes}

    rejection_rates = []
    acceptance_rates = []
    soujorns = {locker_size: 0 for locker_size in locker_sizes}
    for i in range(n_sims):
        locker_dict_copy = copy.deepcopy(locker_dict)
        orders = daily_simulation(day)
        orders.sort(key=lambda x: x[0])

        rejected_orders = []
        for order in orders:
            deposit_time = order[0]
            pickup_time = order[1]
            order_size = order[2]
            order_qty = order[3]

            smallest_size_greater_than_order = [locker_size for locker_size in locker_sizes if locker_size >= order_size][0]
            next_available_size = [locker_size for locker_size in locker_sizes if locker_size >= order_size][1] if len([locker_size for locker_size in locker_sizes if locker_size > order_size]) > 1 else None

            dos = int(pickup_time-deposit_time) if pickup_time >= deposit_time else 24 - deposit_time
            if all(locker_dict_copy[smallest_size_greater_than_order][deposit_time:pickup_time]) >= order_qty:
                locker_dict_copy[smallest_size_greater_than_order][deposit_time:deposit_time+dos] -= np.full(dos, int(order_qty))
                soujorns[smallest_size_greater_than_order] += dos/(n_sims*len(orders))
            
            elif next_available_size and all(locker_dict_copy[next_available_size][deposit_time:pickup_time]) >= order_qty:
                locker_dict_copy[next_available_size][deposit_time:deposit_time+dos] -= np.full(dos, int(order_qty))
                soujorns[next_available_size] += dos/(n_sims*len(orders))

            else: 
                rejected_orders.append(order)
    
        rejection_rate = len(rejected_orders)/len(orders)
        rejection_rates.append(rejection_rate)
        acceptance_rate = 1 - rejection_rate
        acceptance_rates.append(acceptance_rate)
        utilization = {}
        for locker in locker_dict_copy.keys():
            utilization[locker] = np.round(np.mean(np.full(24, locker_dict[locker]) - locker_dict_copy[locker]),2)
    

    return np.round(np.mean(rejection_rates),2), np.round(np.mean(acceptance_rates),2), utilization, soujorns


In [87]:
evaluation(4)

(0.42,
 0.58,
 {1: 2.71, 3: 3.38, 7: 6.21, 10: 3.46, 15: 2.54, 18: 1.79},
 {1: 0.31078861116367584,
  3: 0.5081051957738187,
  7: 1.0002292552396488,
  10: 0.5125552391741093,
  15: 0.39758462949600337,
  18: 0.2097305452611977})

{1: [7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
 3: [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3],
 7: [7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
 10: [4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4],
 15: [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3],
 18: [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]}

In [40]:
[locker_size for locker_size in [1,3,5,7,10,18] if locker_size > 3]

[5, 7, 10, 18]

# Task 4

In [ ]:
min_service_level = 0.99
cost_assigning_d_to_d1_l = 10
c_m = 50
c_s = 50
c_w = 500
# h_max = 15
lamda = 30
# w_max = 20

In [17]:
import gurobipy as gp

m = gp.Model('Task 4')
A = m.addVars(27, 7, vtype=gp.GRB.BINARY, name='A') # 27 rows and 7 column. 1 if that grid is assigned 0 ow




In [65]:
def avg_yearly_demand(df):
    for i in range(1, 6):
        df[f'avg_year_{i}'] = df[f'Year_{i}'] * df['Probability']
        df[f'squared_year_{i}'] = df[f'Year_{i}']**2 * df['Probability']
    return df
annaul_demand_forecast_with_avg = avg_yearly_demand(annaul_demand_forecast)
annaul_demand_forecast_with_avg

Scenario  Probability  Year_1  Year_2  Year_3  Year_4  Year_5  year1_avg  \
0         1         0.30   80000   90000  100000  105000  110000    24000.0   
1         2         0.25   60000   65000   70000   80000   90000    15000.0   
2         3         0.20   80000  100000  110000  120000  150000    16000.0   
3         4         0.15   50000   70000   75000   75000   65000     7500.0   
4         5         0.10   50000   60000   60000   60000   60000     5000.0   

   avg_year_1  avg_year_2  avg_year_3  avg_year_4  avg_year_5  squared_year_1  \
0     24000.0     27000.0     30000.0     31500.0     33000.0    1.920000e+09   
1     15000.0     16250.0     17500.0     20000.0     22500.0    9.000000e+08   
2     16000.0     20000.0     22000.0     24000.0     30000.0    1.280000e+09   
3      7500.0     10500.0     11250.0     11250.0      9750.0    3.750000e+08   
4      5000.0      6000.0      6000.0      6000.0      6000.0    2.500000e+08   

   squared_year_2  squared_year_3  squared_year_4  squared_year_5  
0    2.430000e+09    3.000000e+09    3.307500e+09    3.630000e+09  
1    1.056250e+09    1.225000e+09    1.600000e+09    2.025000e+09  
2    2.000000e+09    2.420000e+09    2.880000e+09    4.500000e+09  
3    7.350000e+08    8.437500e+08    8.437500e+08    6.337500e+08  
4    3.600000e+08    3.600000e+08    3.600000e+08    3.600000e+08

In [60]:
annaul_demand_forecast_with_avg['avg_year_1'].sum(axis=0)

67500.0

In [62]:
def get_avg_yearly_data(df):
    avg_yearly_data = pd.DataFrame(index=['mean', 'std'], columns=['Year_1', 'Year_2', 'Year_3', 'Year_4', 'Year_5'])
    for i in range(1, 6):
        avg_yearly_data.loc['mean', f'Year_{i}'] = df[f'avg_year_{i}'].sum(axis=0)
        avg_yearly_data.loc['std', f'Year_{i}'] = np.sqrt(df[f'squared_year_{i}'].sum(axis=0) - df[f'avg_year_{i}'].sum(axis=0)**2)
    return avg_yearly_data

avg_yearly_data = get_avg_yearly_data(annaul_demand_forecast_with_avg)
avg_yearly_data

Year_1        Year_2        Year_3        Year_4        Year_5
mean       67500.0       79750.0       86750.0       92750.0      101250.0
std   12990.381057  14872.373718  17977.416388  19715.159142  29953.088322

In [64]:
z_99 = 2.326
for i in range(1, 6):
    avg_yearly_data.loc['demand_99', f'Year_{i}'] = avg_yearly_data.loc['mean', f'Year_{i}'] + z_99 * avg_yearly_data.loc['std', f'Year_{i}']

avg_yearly_data

Year_1         Year_2         Year_3         Year_4  \
mean            67500.0        79750.0        86750.0        92750.0   
std        12990.381057   14872.373718   17977.416388   19715.159142   
demand_99  97715.626338  114343.141267  128565.470519  138607.460165   

                  Year_5  
mean            101250.0  
std         29953.088322  
demand_99  170920.883436

In [77]:
from datetime import datetime

def get_demand_instance(year: int, deposit_hour: int, pickup_hour: int, date: str, demand_df = avg_yearly_data, daily_demand = yearly_probs, hourly_demand_df = hourly_demand):
    '''
    Gets the demand instance
    params: year: int, deposit_hour: int, pickup_hour: int, date: Month day
    output: demand_instance: df 
    '''
    date1 = datetime.strptime(date, '%B %d')
    diff = date1 - datetime.strptime('January 1', '%B %d')
    num_days = diff.days
    demand_yearly = demand_df.loc['demand_99', f'Year_{year}']
    demand_daily = demand_yearly*daily_demand.loc[num_days, 'yearly_prob']
    hourly_df = hourly_demand_df*demand_daily

    return hourly_df.iloc[deposit_hour, pickup_hour]

get_demand_instance(4, 10, 20, 'May 1', avg_yearly_data, yearly_probs, hourly_demand)
    

0.29068953282750004

In [ ]:
#bsdbcsdjkbvbkjdvjvn